datalad/interface/utils.py

```diff
@@ -127,83 +132,312 @@ def handle_dirty_datasets(dpaths,
+      no new state will be saved.
+    """
+    # XXX paths must be in the given ds, no further sanity checks!
+
+    # make sure that all pending changes (batched annex operations, etc.)
+    # are actually reflected in Git
+    ds.repo.precommit()
+
+    # track what is to be committed, so it becomes
+    # possible to decide when/what to save further down
+    # and one level up
+    orig_hexsha = ds.repo.get_hexsha()
+
+    # always yields list; empty if None
+    files = list(
+        set(
+            [opj(ds.path, f) if not isabs(f) else f for f in assure_list(paths)]))
+
+    # try to consider existing and changed files, and prevent untracked
+    # files from being added
+    # XXX not acting upon untracked files would be very expensive, because
+    # I see no way to avoid using `add` below and git annex has no equivalent
+    # to git add's --update -- so for now don't bother
+    # XXX alternatively we could consider --no-ignore-removal to also
+    # have it delete any already vanished files
+    # asking yourself why we need to `add` at all? For example, freshly
+    # unlocked files in a v5 repo are listed as "typechange" and commit
+    # refuses to touch them without an explicit `add`
+    tostage = [f for f in files if lexists(f)]
+    if tostage:
+        lgr.debug('staging files for commit: %s', tostage)
+        if isinstance(ds.repo, AnnexRepo):
+            # to make this work without calling `git add` in addition,
+            # this needs git-annex v6.20161210 (see #1027)
+            ds.repo.add(tostage, commit=False)
+        else:
+            # --update will ignore any untracked files, sadly git-annex add
+            # above does not
+            # will complain about vanished files though, filter them here, but
+            # keep them for a later commit call
+            ds.repo.add(tostage, git_options=['--update'], commit=False)
+
+    _datalad_msg = False
+    if not message:
+        message = 'Recorded existing changes'
+        _datalad_msg = True
+
+    if files or ds.repo.repo.is_dirty(
+            index=True,
+            working_tree=False,
+            untracked_files=False,
+            submodules=True):
+        # either we have an explicit list of files, or we have something
+        # stages otherwise do not attempt to commit, as the underlying
+        # repo will happily commit any non-change
+        # not checking the working tree or untracked files should make this
+        # relavtively cheap
+
+        # TODO: commit() should rather report a dedicated ValueError
+        # waiting for #1170
+        from datalad.support.exceptions import CommandError
+        try:
+            # we will blindly call commit not knowing if there is anything to
+            # commit -- this is cheaper than to anticipate all possible ways
+            # a repo in whatever mode is dirty
+            # however, if nothing is dirty the whining wil start
+            # --> sucking it up right here
```

Comment:
Would you mind making it an issue to move this processing of `CommandError` to `GitRepo.commit`?